### PR TITLE
feat: display accurate copyright years on pages

### PR DESCRIPTION
**Pull request by jamesericdavidson**
tl;dr Implementation for #222.

- Pages which are *not* of kind `page` continue to use `now.Year`
- Pages which do *not* declare `.Date` also use `now.Year`
- Pages which declare `.Date` use it *instead* of `now.Year`
- Pages which declare `.Lastmod` and `.Date` use both dates in a range, *instead* of `now.Year`

A long `copyright` string will overflow the `.footer` container, ergo some assistance on the CSS side would be greatly appreciated. :smiley: 

N.B. Copyright is described by Hugo here: https://gohugo.io/methods/site/copyright/

- [ ] css fixes
- [x] docs (advanced)
- [x] screenshots
-  ~~config opt-in *?*~~
   ~~`copyright` is only displayed if it exists in `config.toml`; I'd argue this isn't necessary~~
- ~~add `dateFormat` to `config.toml` *?*~~
   - ~~add `Lastmod` to any relevant `exampleSite` pages for illustration *?*~~

### CODE REVIEW

**Comment by jamesericdavidson** [comment]
# Screenshots

Non-page kinds:

image

Page without `.Date` or `.Lastmod`:

image

Pages with `.Date` but not `.Lastmod`:

image

Pages with `.Date` and `.Lastmod`:

image

Sites with `copyright`:

image

**Comment by jamesericdavidson** [comment]
@yashmehrotra @526avijitgupta I hope this suits. Look forward to your review :+1:

File: layouts/partials/footer.html [hunk 1]
```diff
@@ -4,9 +4,17 @@
...
-    <span>&copy; {{ now.Year }} {{ .Site.Params.Footer}}</span>
+    {{/* Print a year range if .Date and .Lastmod are present */}}
+    {{ if and (.Date) (.Lastmod) (.Site.Params.Footer) (eq .Kind "page") (gt (time.Format "2006" .Lastmod) (time.Format "2006" .Date)) }}
+    <span>&copy; {{ time.Format "2006" .Date }}-{{ time.Format "2006" .Lastmod }} {{ .Site.Params.Footer }} {{ .Site.Copyright }}</span>
+    {{/* Use .Date as the copyright year */}}
+    {{ else if and (.Date) (.Site.Params.Footer) (eq .Kind "page") }}
+        <span>&copy; {{ time.Format "2006" .Date }} {{ .Site.Params.Footer }} {{ .Site.Copyright }}</span>
+    {{/* Use the current year for other Kinds, and pages without Date or Lastmod */}}
+    {{ else if (.Site.Params.Footer) }}
+        <span>&copy; {{ now.Year }} {{ .Site.Params.Footer }}</span>
```
Comment (yashmehrotra) [inline layouts/partials/footer.html:15]: Can you change the logic flow to
```template
{{ footerText := "&copy; now.year" }}
// logic for setting copyRight in footerText

{{ if .params.footer}}
  footerText += " {{footer}""
  
<span>{{ footertext }}</span>
```

**Review by 526avijitgupta — APPROVED** [review]
@jamesericdavidson LGTM. Thanks for the fix and for providing screenshots in the PR :)

@yashmehrotra Please merge once it looks okay from your side

File: layouts/partials/footer.html (hunk 1 above)
Comment (jamesericdavidson) [reply layouts/partials/footer.html:15]: Thanks @yashmehrotra,

I tried something similar to your syntax, but `+=` is not a supported operator. [Using variables](https://gohugo.io/templates/partial/#variable-scoping) didn't provide much joy either.

First example (code fails):

```
    {{ $footerText := "" }}

    {{/* Print a year range if .Date and .Lastmod are present */}}
    {{ if and (.Date) (.Lastmod) (.Site.Params.Footer) (eq .Kind "page") (gt (time.Format "2006" .Lastmod) (time.Format "2006" .Date)) }}
        {{ $footerText := "&copy; {{ time.Format '2006' .Date }}-{{ time.Format '2006' .Lastmod }} {{ .Site.Params.Footer }} {{ .Site.Copyright }}" }}

    {{/* Use .Date as the copyright year */}}
    {{ else if and (.Date) (.Site.Params.Footer) (eq .Kind "page") }}
        {{ $footerText := "&copy; {{ time.Format '2006' .Date }} {{ .Site.Params.Footer }} {{ .Site.Copyright }}" }}

    {{/* Use the current year for other Kinds, and pages without Date or Lastmod */}}
    {{ else if (.Site.Params.Footer) }}
        {{ $footerText := "&copy; {{ now.Year }} {{ .Site.Params.Footer }}" }}
    {{ end }}

    <span>{{ $footerText }}</span>
```

After some reading, I tried [Scratch](https://gohugo.io/methods/page/scratch/#methods), however this produces even longer code. `.Scratch.Add` also suffers with arbitrary characters such as a space `" "` or dash `"-"`.

Second example (code works, but there is no formatting. See below screenshot):

```
    {{ .Scratch.Set "footerText" "" }}

    {{/* Print a year range if .Date and .Lastmod are present */}}
    {{ if and (.Date) (.Lastmod) (.Site.Params.Footer) (eq .Kind "page") (gt (time.Format "2006" .Lastmod) (time.Format "2006" .Date)) }}
        {{ .Scratch.Add "footerText" (.Date | time.Format "2006") }}
        {{ .Scratch.Add "footerText" (.Lastmod | time.Format "2006") }}
        {{ .Scratch.Add "footerText" .Site.Params.Footer }}
        {{ .Scratch.Add "footerText" .Site.Copyright }}

    {{/* Use .Date as the copyright year */}}
    {{ else if and (.Date) (.Site.Params.Footer) (eq .Kind "page") }}
        {{ .Scratch.Add "footerText" (.Date | time.Format "2006") }}
        {{ .Scratch.Add "footerText" .Site.Params.Footer }}
        {{ .Scratch.Add "footerText" .Site.Copyright }}

    {{/* Use the current year for other Kinds, and pages without Date or Lastmod */}}
    {{ else if (.Site.Params.Footer) }}
        {{ .Scratch.Add "footerText" (time.Now | time.Format "2006") }}
        {{ .Scratch.Add "footerText" .Site.Params.Footer }}
    {{ end }}
```

image

N.B. Don't ask why `.Scratch.Add` is happier with invoking `time.Format` after a bar character. I don't know either :grin: 

More research required. Just keeping you up-to-date.

Comment (jamesericdavidson) [reply layouts/partials/footer.html]: 6268dd9 After some trial and error, I believe I've compressed this down as much as possible. Hope this is what you intended :+1:

**Comment by jamesericdavidson** [comment]
Highlighting the CSS problems arising, as mentioned in https://github.com/526avijitgupta/gokarna/pull/233#issue-2540425501:

- Long footer overflows with Table of Contents
- Long footer overflows its container and bisects content

# Screenshots

image

image

**Comment by yashmehrotra** [comment]
@jamesericdavidson I think we can leave the CSS problem for now since it only happens with very large texts which is not the usual case. Can you create a separate issue for this ?

I'll merge this for now, thanks for your contribution

**Comment by jamesericdavidson** [comment]
@yashmehrotra Brilliant, thanks Yash. Will do :+1:
